src/softmax.jl

```diff
@@ -94,14 +98,18 @@ logsoftmax(x; dims = 1) = logsoftmax!(similar(x, (float ∘ eltype)(x)), x; dims
 
 logsoftmax!(x; dims = 1) = logsoftmax!(x, x; dims = dims)
 
-function logsoftmax!(out::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
-    out .= x .- maximum(x; dims = dims)
-    # out .-= log.(sum(exp.(out); dims = dims))  # WARN: this will decrease performance.
+function logsoftmax!(out::AbstractArray{T}, x::AbstractArray; dims = 1) where {T}
+    max_ = maximum(x; dims = dims)
+    if all(isfinite, max_)
+        out .= x .- max_
+    else
+        @. out = ifelse(isequal(max_,Inf), ifelse(isequal(x,Inf), 0, -Inf), x - max_)
+    end
```

Comment:
why you are using a slightly different pattern here compared to the softmax! change?

Reply:
I don't remember exactly. In the case which does have `Inf`, this version does one pass over `out`, while the other does two. Will check that making `softmax!` do the one-pass version doesn't hurt anything.

Reply:
Now changed to match.